File: src/org/elixir_lang/annotator/Parameter.java
package org.elixir_lang.annotator;

import com.intellij.psi.NavigatablePsiElement;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import org.elixir_lang.psi.*;
import org.elixir_lang.psi.call.Call;
import org.elixir_lang.structure_view.element.Delegation;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class Parameter {
    public enum Type {
        FUNCTION_NAME,
        MACRO_NAME,
        VARIABLE;

        public static boolean isCallDefinitionClauseName(Type type) {
            return type == FUNCTION_NAME || type == MACRO_NAME;
        }
    }

    /*
     * Public Static Methods
     */

    /**
     * A new {@link Parameter} with {@link #parameterized} filled in if {@code parameter}'s {@link #entrance} is a
     * parameter element.
     *
     * @return a new {@link Parameter} with {@link #parameterized} filled in if {@link #entrance} is a valida parameter
     *   element.
     */
    @Contract(pure = true)
    @NotNull
    public static Parameter putParameterized(final @NotNull Parameter parameter) {
        return putParameterized(parameter, parameter.entrance);
    }

    /*
     * Private Static Methods
     */

    @Contract(pure = true)
    @NotNull
    private static <T> T notNullize(@Nullable T nullable, @NotNull T defaultValue) {
        T notNull;

        if (nullable == null) {
            notNull = defaultValue;
        } else {
            notNull = nullable;
        }

        return notNull;
    }

    @Contract(pure = true)
    @NotNull
    private static Parameter putParameterized(@NotNull final Parameter parameter, final @NotNull Call ancestor) {
        Parameter parameterizedParameter;

        if (CallDefinitionClause.isFunction(ancestor) || Delegation.is(ancestor)) {
            parameterizedParameter = new Parameter(
                    parameter.defaultValue,
                    parameter.entrance,
                    notNullize(parameter.parameterized, ancestor),
                    notNullize(parameter.type, Type.FUNCTION_NAME)
            );
        } else if (CallDefinitionClause.isMacro(ancestor)) {
            parameterizedParameter = new Parameter(
                    parameter.defaultValue,
                    parameter.entrance,
                    notNullize(parameter.parameterized, ancestor),
                    notNullize(parameter.type, Type.MACRO_NAME)
            );
        } else if (ancestor.hasDoBlockOrKeyword()) {
            parameterizedParameter = new Parameter(
                    parameter.defaultValue,
                    parameter.entrance,
                    ancestor,
                    notNullize(parameter.type, Type.VARIABLE)
            );
        } else {
            PsiElement element = ancestor.functionNameElement();
            Parameter updatedParameter = parameter;

            if (!PsiTreeUtil.isAncestor(element, parameter.entrance, false)) {
                updatedParameter = new Parameter(
                        parameter.defaultValue,
                        parameter.entrance,
                        ancestor,
                        notNullize(parameter.type, Type.VARIABLE)
                );
            }

            // use generic handling so that parent is checked
            parameterizedParameter = putParameterized(updatedParameter, (PsiElement) ancestor);
        }

        return parameterizedParameter;
    }

    @Contract(pure = true)
    @NotNull
    private static Parameter putParameterized(@NotNull final Parameter parameter,
                                              @NotNull final ElixirAnonymousFunction ancestor) {
        return new Parameter(
                parameter.defaultValue,
                parameter.entrance,
                ancestor,
                notNullize(parameter.type, Type.VARIABLE)
        );
    }

    @Contract(pure = true)
    @NotNull
    private static Parameter putParameterized(@NotNull final Parameter parameter, @NotNull final PsiElement ancestor) {
        PsiElement parent = ancestor.getParent();

        if (parent == null) {
            return new Parameter(parameter.entrance);
        }

        return switch (ParameterWalk.classify(parent)) {
            case RECURSE -> putParameterized(parameter, parent);
            case CALL -> putParameterized(parameter, (Call) parent);
            case ANONYMOUS_FUNCTION -> putParameterized(parameter, (ElixirAnonymousFunction) parent);
            case STOP, LEAF -> new Parameter(parameter.entrance);
        };
    }

    /*
     * Fields
     */

    @Nullable
    public final PsiElement defaultValue;
    @NotNull
    public final PsiElement entrance;
    @Nullable
    public final NavigatablePsiElement parameterized;
    @Nullable
    public final Type type;

    /*
     * Constructors
     */

    public Parameter(@NotNull PsiElement entrance) {
        this.defaultValue = null;
        this.entrance = entrance;
        this.parameterized = null;
        this.type = null;
    }

    private Parameter(@Nullable PsiElement defaultValue,
                      @NotNull PsiElement entrance,
                      @Nullable NavigatablePsiElement parameterized,
                      @Nullable Type type) {
        this.defaultValue = defaultValue;
        this.entrance = entrance;
        this.parameterized = parameterized;
        this.type = type;
    }

    /*
     * Public Instance Methods
     */

    /**
     * Whether the {@link #type} is call definition clause name
     *
     * @return {@code true} if {@link #type} is {@link Type#FUNCTION_NAME} or {@link Type#MACRO_NAME}.
     */
    @Contract(pure = true)
    public boolean isCallDefinitionClauseName() {
        return Type.isCallDefinitionClauseName(type);
    }

    /**
     * Whether {@link #entrance} represents a parameter to a {@link #parameterized} element
     * @return {@code true} if {@link #parameterized} is not {@code null}
     */
    @Contract(pure = true)
    boolean isValid() {
        return parameterized == null;
    }

    /**
     * A Parameter that is not a parameter to anything.
     *
     * @param parameter The original {@link Parameter} that may or may not be parameterized
     * @return an invalid parameter
     */
    @NotNull
    public Parameter not(final @NotNull Parameter parameter) {
        Parameter not;

        if (parameter.defaultValue == null && parameter.parameterized == null) {
            not = parameter;
        } else {
            not = new Parameter(parameter.entrance);
        }

        return not;
    }
}

File: src/org/elixir_lang/annotator/Parameter.kt
package org.elixir_lang.annotator

import com.intellij.psi.NavigatablePsiElement
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import org.elixir_lang.errorreport.Logger.error
import org.elixir_lang.psi.*
import org.elixir_lang.psi.call.Call
import org.elixir_lang.psi.operation.InMatch
import org.elixir_lang.psi.operation.When
import org.elixir_lang.semantic.call.definition.clause.Time
import org.elixir_lang.semantic.semantic
import org.jetbrains.annotations.Contract

class Parameter {
    enum class Type {
        FUNCTION_NAME, MACRO_NAME, VARIABLE;

        companion object {
            fun isCallDefinitionClauseName(type: Type?): Boolean = type == FUNCTION_NAME || type == MACRO_NAME
        }
    }

    val defaultValue: PsiElement?
    val entrance: PsiElement
    val parameterized: NavigatablePsiElement?

    @JvmField
    val type: Type?

    constructor(entrance: PsiElement) {
        defaultValue = null
        this.entrance = entrance
        parameterized = null
        type = null
    }

    private constructor(
        defaultValue: PsiElement?,
        entrance: PsiElement,
        parameterized: NavigatablePsiElement?,
        type: Type?
    ) {
        this.defaultValue = defaultValue
        this.entrance = entrance
        this.parameterized = parameterized
        this.type = type
    }

    /**
     * Whether the [.type] is call definition clause name
     *
     * @return `true` if [.type] is [Type.FUNCTION_NAME] or [Type.MACRO_NAME].
     */
    @get:Contract(pure = true)
    val isCallDefinitionClauseName: Boolean
        get() = Type.isCallDefinitionClauseName(type)

    /**
     * Whether [.entrance] represents a parameter to a [.parameterized] element
     * @return `true` if [.parameterized] is not `null`
     */
    @get:Contract(pure = true)
    val isValid: Boolean
        get() = parameterized == null

    /**
     * A Parameter that is not a parameter to anything.
     *
     * @param parameter The original [Parameter] that may or may not be parameterized
     * @return an invalid parameter
     */
    fun not(parameter: Parameter): Parameter = if (parameter.defaultValue == null && parameter.parameterized == null) {
        parameter
    } else {
        Parameter(parameter.entrance)
    }

    companion object {
        /**
         * A new [Parameter] with [.parameterized] filled in if `parameter`'s [.entrance] is a
         * parameter element.
         *
         * @return a new [Parameter] with [.parameterized] filled in if [.entrance] is a valida parameter
         * element.
         */
        @Contract(pure = true)
        fun putParameterized(parameter: Parameter): Parameter = putParameterized(parameter, parameter.entrance)


        private fun error(message: String, element: PsiElement) {
            error(Parameter::class.java, message + " (when element class is " + element.javaClass.name + ")", element)
        }

        @Contract(pure = true)
        private fun putParameterized(parameter: Parameter, ancestor: Call): Parameter {
            val parameterizedParameter: Parameter
            val ancestorSemantic = ancestor.semantic

            if ((ancestorSemantic is org.elixir_lang.semantic.call.definition.Clause &&
                        ancestorSemantic.definition.time == Time.RUN) ||
                ancestorSemantic is org.elixir_lang.semantic.call.definition.Delegation
            ) {
                parameterizedParameter = Parameter(
                    parameter.defaultValue,
                    parameter.entrance,
                    parameter.parameterized ?: ancestor,
                    parameter.type ?: Type.FUNCTION_NAME
                )
            } else if (ancestorSemantic is org.elixir_lang.semantic.call.definition.Clause &&
                ancestorSemantic.definition.time == Time.COMPILE
            ) {
                parameterizedParameter = Parameter(
                    parameter.defaultValue,
                    parameter.entrance,
                    parameter.parameterized ?: ancestor,
                    parameter.type ?: Type.MACRO_NAME
                )
            } else if (ancestor.hasDoBlockOrKeyword()) {
                parameterizedParameter = Parameter(
                    parameter.defaultValue,
                    parameter.entrance,
                    ancestor,
                    parameter.type ?: Type.VARIABLE
                )
            } else {
                val element = ancestor.functionNameElement()
                var updatedParameter = parameter
                if (!PsiTreeUtil.isAncestor(element, parameter.entrance, false)) {
                    updatedParameter = Parameter(
                        parameter.defaultValue,
                        parameter.entrance,
                        ancestor,
                        parameter.type ?: Type.VARIABLE
                    )
                }

                // use generic handling so that parent is checked
                parameterizedParameter = putParameterized(updatedParameter, ancestor as PsiElement)
            }
            return parameterizedParameter
        }

        @Contract(pure = true)
        private fun putParameterized(
            parameter: Parameter,
            ancestor: ElixirAnonymousFunction
        ): Parameter = Parameter(
            parameter.defaultValue,
            parameter.entrance,
            ancestor,
            parameter.type ?: Type.VARIABLE
        )

        @Contract(pure = true)
        private fun putParameterized(parameter: Parameter, ancestor: PsiElement): Parameter =
            when (val parent = ancestor.parent) {
                is When -> {
                    putParameterized(parameter, parent)
                }
                is Call -> {
                    putParameterized(
                        parameter,
                        parent
                    )
                }
                is AtOperation, is ElixirAccessExpression, is ElixirAssociations, is ElixirAssociationsBase,
                is ElixirEex, is ElixirEexTag, is ElixirBitString, is ElixirBracketArguments,
                is ElixirContainerAssociationOperation, is ElixirKeywordPair, is ElixirKeywords, is ElixirList,
                is ElixirMapArguments, is ElixirMapConstructionArguments, is ElixirMapOperation,
                is ElixirMatchedParenthesesArguments, is ElixirNoParenthesesArguments,
                is ElixirNoParenthesesKeywordPair, is ElixirNoParenthesesKeywords,
                is ElixirNoParenthesesManyStrictNoParenthesesExpression, is ElixirNoParenthesesOneArgument,
                is ElixirNoParenthesesStrict, is ElixirParenthesesArguments, is ElixirParentheticalStab, is ElixirStab,
                is ElixirStabNoParenthesesSignature, is ElixirStabBody, is ElixirStabOperation,
                is ElixirStructOperation, is ElixirTuple -> {
                    putParameterized(parameter, parent)
                }
                is ElixirAnonymousFunction -> {
                    putParameterized(parameter, parent)
                }
                is InMatch -> {
                    putParameterized(parameter, parent)
                }
                is AtUnqualifiedBracketOperation, is BracketOperation, is ElixirBlockItem, is ElixirDoBlock,
                is ElixirInterpolation, is ElixirMapUpdateArguments, is ElixirMultipleAliases, is ElixirHeredocLineBody,
                is ElixirLineBody, is PsiFile, is QualifiedAlias, is QualifiedMultipleAliases -> {
                    Parameter(parameter.entrance)
                }
                else -> {
                    error("Don't know how to check if parameter", parent)
                    Parameter(parameter.entrance)
                }
            }
    }
}
